src/ensemble_scheduler.cc

```diff
@@ -774,11 +781,13 @@ EnsembleContext::PrepareSteps(
         ensemble_status_ = GetNextSteps(updated_tensors, ready_steps);
       }
 
-      // Check and send ensemble response
-      if ((!ensemble_status_.IsOk()) || (inflight_step_counter_ == 0) ||
-          info_->is_decoupled_) {
+      // Reaching the end of processing completed step, check if the ensemble
+      // should send response / error.
+      if (!ensemble_status_.IsOk()) {
```

Comment:
I see, so the previous logic was checking `ensemble_status_`, `inflight_step_counter_` and `info_->is_decoupled_ ` here. The new logic introduced a more precise behavior based on  `inflight_step_counter_` and `info_->is_decoupled_  values in combination with `response_sent_`.

Reply:
Yes, previously the conditional will "trim" the number of possible ensemble responses, which causes the ensemble response not being filled properly.